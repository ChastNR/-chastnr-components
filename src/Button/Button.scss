.btn {
  border-radius: 8px;
  cursor: pointer;
  position: relative;
  transition-duration: 0.25s;
  transition-property: background-color, opacity, border-color, box-shadow;
  transition-timing-function: ease;
  user-select: none;

  .btn__txt__cntr {
    align-items: center;
    display: flex;
    flex-direction: row;
    flex-grow: 1;
    justify-content: center;
    min-height: 42px;
    min-width: 32px;
    outline: 0;
    padding: 9px 23px 7px;
    position: relative;

    .btn__txt__cntr__txt {
      font-size: 15px;
      font-weight: 400;
      letter-spacing: 0;
      line-height: 24px;
      word-wrap: break-word;
    }
  }

  &.btn--link {
    background: transparent;
    border: none;
    color: #336fee;

    &:hover:not([disabled]) {
      background-color: #e4ebf3;
      border-color: #e4ebf3;
    }
  }

  &.btn--icon {
    background: transparent;
    border: none;
    color: #336fee;
  }

  &.btn--outline {
    background-color: transparent;
    border: 1px solid #336fee;
    color: #336fee;

    &:hover:not([disabled]) {
      background-color: rgba(236, 241, 247, 0.08);
      border-color: #1f50ba;
    }
  }

  background-color: var(--blue);
  border: 1px solid var(--blue);
  color: #fff;

  &:hover:not([disabled]) {
    background-color: var(--dark-blue);
    border-color: var(--dark-blue);
  }

  &:disabled {
    cursor: default;
    opacity: 0.56;
  }
}
