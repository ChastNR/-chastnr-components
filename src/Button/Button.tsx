import { Link } from 'react-router-dom';

import classNames from 'clsx';

import { Ring } from '../base';

import './Button.scss';

interface LocationProps {
  pathname?: string;
  search?: string;
  state?: unknown;
}

interface ButtonProps {
  buttonStyle?: 'link' | 'outline' | 'icon';
  className?: string;
  disabled?: boolean;
  isLoading?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  to?: string | LocationProps;
  type?: 'submit' | 'reset' | 'button';
}

const Button: React.FC<ButtonProps> = ({
  buttonStyle,
  children,
  className,
  disabled,
  isLoading,
  onClick,
  to,
  type = 'button',
}) => {
  const buttonClassNames = classNames('btn', className, {
    btn__link: buttonStyle === 'link',
    btn__icon: buttonStyle === 'icon',
    btn__outline: buttonStyle === 'outline',
  });

  const button = (
    <button className={buttonClassNames} disabled={disabled} onClick={onClick} type={type}>
      <div className="btn__txt__cntr" tabIndex={-1}>
        <div className="btn__txt__cntr__txt">{isLoading ? <Ring /> : children}</div>
      </div>
    </button>
  );

  return to ? (
    <Link tabIndex={-1} to={to}>
      {button}
    </Link>
  ) : (
    button
  );
};

export default Button;
