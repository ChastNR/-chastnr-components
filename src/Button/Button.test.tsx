import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import Button from "./Button";

describe("<Button />", () => {
  it("should match snapshot", () => {
    const button = shallow(<Button>Click</Button>);
    expect(toJson(button)).toMatchSnapshot();
  });

  it("props should be equal", () => {
    const style = "outline";
    const className = "testClassName";
    const disabled = true;
    const isLoading = true;
    const to = "";
    const type = "submit";

    const button = mount(
      <Button
        buttonStyle={style}
        className={className}
        disabled={disabled}
        isLoading={isLoading}
        to={to}
        type={type}
      />
    );

    const props = button.props();

    expect(props.buttonStyle).toEqual(style);
    expect(props.className).toEqual(className);
    expect(props.disabled).toEqual(disabled);
    expect(props.isLoading).toEqual(isLoading);
    expect(props.to).toEqual(to);
    expect(props.type).toEqual(type);
  });
});
