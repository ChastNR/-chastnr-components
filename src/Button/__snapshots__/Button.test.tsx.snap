// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
<button
  class="btn testClassName btn--outline"
  disabled=""
  type="submit"
>
  <div
    class="btn__txt__cntr"
    tabindex="-1"
  >
    <div
      class="btn__txt__cntr__txt"
    >
      <div
        class="ring ring--small"
      >
        <span>
          content
        </span>
      </div>
    </div>
  </div>
</button>
`;
